File: LintCode/Easy/453. Flatten Binary Tree to Linked List/Solution.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

public class Solution {
    /**
     * @param root: a TreeNode, the root of the binary tree
     * @return: nothing
     */

    private TreeNode prevNode = null;
    
    public void flatten(TreeNode root) {
        // write your code here
        
        if (root == null) return;
        
        // last node you traversed
        if (prevNode != null) {
            prevNode.left = null;
            prevNode.right = root;

        }
        
        // The node just traversed
        prevNode = root;
        
        // keep the right subtree
        TreeNode currRight = root.right;
        
        // traverse subtree
        flatten(root.left);
        flatten(currRight);

    }

}

File: LintCode/Easy/453. Flatten Binary Tree to Linked List/Solution.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: a TreeNode, the root of the binary tree
    @return: nothing
    """
    def flatten(self, root):
        # write your code here
        self.flattenHelper(root)


    def flattenHelper(self, root):
        if not root:
            return root

        left  = self.flattenHelper(root.left)
        right = self.flattenHelper(root.right)

        # Re-connect leftmost leaf to root.right
        if left:
            left.right = root.right
            root.right = root.left
            root.left = None

        # Return rightmost leaf for next recursion
        if right:
            return right

        # Return leftmost if there is no right subtree
        if left:
            return left

        # Reached leaf node, return leaf
        return root
